### PR TITLE
Added bind() tests

### PR DESCRIPTION
**Pull request by stevenengler**
These tests do not include the implicit bind tests of `src/test/bind/test_bind.c` since they do not test the `bind()` call. Instead they should later be implemented in a different test directory.

### CODE REVIEW

**Comment by robgjansen** [comment]
I'm noticing in the codecov reports that coverage is being shown of the test code itself (`test/socket/bind/test_bind.rs`) in addition to the shadow code it is testing (`main/host/syscall/socket.c`). Is this the standard way of showing coverage, or should we be ignoring how much of the test code we are covering?

**Comment by stevenengler** [comment]
> I'm noticing in the codecov reports that coverage is being shown of the test code itself (`test/socket/bind/test_bind.rs`) in addition to the shadow code it is testing (`main/host/syscall/socket.c`). Is this the standard way of showing coverage, or should we be ignoring how much of the test code we are covering?

I noticed this too (we're processing the coverage of all tests), and while I don't think displaying coverage of the tests themselves is common, I don't personally think there's a harm in doing so (more information is better than less). But it's probably best to remove coverage of the tests themselves just to be more like other projects.

**Comment by sporksmith** [comment]
FWIW I like having the coverage of the tests themselves, since it helps identify dead code there (and helped me catch when I'd accidentally left a test disabled in a previous PR :))

File: src/test/socket/bind/test_bind.rs [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+    args: &BindArguments,
+    should_error: bool,
+    expected_errno: Option<libc::c_int>,
+) -> Result<(), String> {
+    // get a pointer to the sockaddr and the size of the structure
+    // careful use of references here makes sure we don't copy memory, leading to stale pointers
+    let (addr_ptr, real_addr_len) = match &args.addr {
+        Some(addr) => match &addr {
+            &LibcSockAddr::In(sockaddr_in) => (
+                sockaddr_in as *const _ as *const libc::sockaddr,
```
Comment (sporksmith) [inline src/test/socket/bind/test_bind.rs:534]: I'm having trouble following this. Could you walk me through it a bit? (Might be worth a comment, but OTOH might not if it's just a matter of me not knowing the language well enough yet)

File: src/test/socket/bind/test_bind.rs [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+
+// test binding using an argument that could be a fd, but is not
+fn test_non_existent_fd() -> Result<(), String> {
+    let args = BindArguments {
+        fd: 8934,
+        addr: None,
+        addr_len: Some(5),
+    };
+
+    check_bind_call(&args, true, Some(libc::EBADF))?;
```
Comment (sporksmith) [inline src/test/socket/bind/test_bind.rs:184]: For most/all of these: isn't using the `?` operator here followed by returning `Ok(())` equivalent to just returning this result?  e.g.
```
  ...
  check_bind_call(&args, true, Some(libc::EBADF))
}
```

File: src/test/socket/bind/test_bind.rs [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let args = BindArguments {
+        fd: fd,
+        addr: Some(LibcSockAddr::In(addr)),
+        addr_len: Some(5),
```
Comment (sporksmith) [inline src/test/socket/bind/test_bind.rs:235]: Maybe clearer as `size_of<libc::sockaddr_in>-1` instead of `5`?

File: src/test/socket/bind/test_bind.rs [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+    // make sure we didn't have any duplicate tests
+    assert_eq!(num_tests, tests.len());
+
+    // add all of the passing tests
+    tests.extend(get_passing_tests());
+
+    tests
+}
+
+fn run_tests<'a, I>(tests: I, summarize: bool) -> Result<(), ()>
```
[the remark this answers is not in the document]
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs:140]: Possibly, but I'm waiting to see how future modules are organized to figure out the best way to do this. For example there is a similar function in the `test_socket.rs` code, but it's different enough that there isn't a good way to generalize it.

File: src/test/socket/bind/test_bind.rs [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+    ];
+
+    // tests to repeat for different socket options
+    for &sock_type in [libc::SOCK_STREAM, libc::SOCK_DGRAM].iter() {
+        for &flag in [0, libc::SOCK_NONBLOCK, libc::SOCK_CLOEXEC].iter() {
+            #[rustfmt::skip]
+            let mut more_tests: Vec<(String, TestFn)> = vec![
+                // Docker does not support IPv6, so this test is disabled for now
+                //("test_ipv6".to_string(),
+                //    Box::new(move || test_ipv6(sock_type, flag)))
```
[the remark this answers is not in the document]
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs:116]: They're expected to fail with Shadow, but we run them outside of Shadow where they're expected to pass. But they actually fail outside of Shadow when run in Docker.

File: src/test/socket/bind/test_bind.rs [hunk 6]
```diff
@@ -0,0 +1,588 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+// Docker does not support IPv6, so IPv6 is not currently used
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs:7]: Rust considers the `In6(libc::sockaddr_in6)` within the enum to be dead code, and I don't think there's a way to specify that the single variant is dead-code.

File: src/test/socket/bind/test_bind.rs [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+    unsafe {
+        let error_ptr = libc::strerror(errno);
+        cstr = std::ffi::CStr::from_ptr(error_ptr)
+    }
+    cstr.to_string_lossy().into_owned()
+}
+
+fn check_bind_call(
+    args: &BindArguments,
+    should_error: bool,
```
[the remark this answers is not in the document]
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs:526]: I originally made it this way since you might want to say that something fails, but you don't care about the errno. I did this with the expectation that different versions of glibc might have changed the errno at some point (since some of the errnos, especially when there are multiple different errors, are arbitrary), but in practice this didn't seem to be the case for the socket() or bind() calls. I'll remove this from these tests.

File: src/test/socket/bind/test_bind.rs (hunk 2 above)
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs:184]: Good catch! I missed that.

File: src/test/socket/bind/test_bind.rs (hunk 1 above)
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs:534]: I added some comments here to hopefully make it clearer, and I replaced the `_` with an explicit type. But `bind()` requires the underlying address type (ex: `libc::sockaddr_in` or `libc::sockaddr_in6`) to be cast to a `libc::sockaddr`. So the `sockaddr_in as *const _` casts the reference of type `&libc::sockaddr_in` to a pointer `*const libc::sockaddr_in`, and the `as *const libc::sockaddr` switches the pointer type from `*const libc::sockaddr_in` to `*const libc::sockaddr`. Rust requires the two casts since you cannot cast a reference (`&sockaddr_in`) directly to a pointer of a different type (`*const libc::sockaddr`).

File: src/test/socket/bind/test_bind.rs [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+        ("test_short_addr".to_string(),
+            Box::new(test_short_addr)),
+    ];
+
+    // tests to repeat for different socket options
+    for &sock_type in [libc::SOCK_STREAM, libc::SOCK_DGRAM].iter() {
+        for &flag in [0, libc::SOCK_NONBLOCK, libc::SOCK_CLOEXEC].iter() {
+            #[rustfmt::skip]
+            let mut more_tests: Vec<(String, TestFn)> = vec![
+                ("test_ipv4".to_string(),
```
[the remark this answers is not in the document]
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs]: Done

File: src/test/socket/bind/test_bind.rs [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+            #[rustfmt::skip]
+            let mut more_tests: Vec<(String, TestFn)> = vec![
+                // Docker does not support IPv6, so this test is disabled for now
+                //("test_ipv6".to_string(),
+                //    Box::new(move || test_ipv6(sock_type, flag)))
+            ];
+
+            // add details to the test names to avoid duplicates
+            for (test_name, _) in more_tests.iter_mut() {
+                *test_name = format!("{} <type={},flag={}>", test_name, sock_type, flag);
```
[the remark this answers is not in the document]
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs]: Done

File: src/test/socket/bind/test_bind.rs (hunk 3 above)
Comment (stevenengler) [reply src/test/socket/bind/test_bind.rs]: Done
